opt: make outside-of-histogram estimates more pessimistic

DrewKimball opened the pull request:
#### opt: improve outside-histogram optimizer tests

This commit makes the following improvements to the `outside-histogram`
optimizer test:
* Fixes a minor mistake in the test, where the upper bounds on a histogram
  for a unique column were incorrect.
* Simplifies the test code for setting session variable defaults.
* Adds comments explaining each test case.

Epic: None

Release note: None

#### opt: add unique and distinct column to outside-histogram test

This commit adds a column `u` with a unique index, and a column `d`
with a non-unique index to the outside-histogram test table. Both
columns use the same histogram as the primary key `k` (and therefore
have no duplicate values).

This commit also adds tests that are similar to `Q1`, but using the
new columns instead of `k`. This shows the optimizer's behavior when
the "good" plan requires an index join, and when filtering on `d`,
shows how the optimizer behaves when no plan limits the max cardinality
of the query.

Epic: None

Release note: None

#### opt: add outside-histogram test for missing enum value

This commit adds a pair of `outside-histogram` optimizer test cases that
simulate a query against an enum column for which one of the values was not
sampled.

Epic: None

Release note: None

#### opt: add outside-histogram tests with large table

This commit adds to the existing optimizer tests for filters that select
values outside of histograms. The new tests are against a large table
to show how the optimizer's trust of low-selectivity estimates varies
with table size.

Epic: None

Release note: None

#### opt: make outside-of-histogram estimates more pessimistic

This commit makes rowcount estimation fall back on distinct count estimates
when a constraint includes zero histogram values. This biases the optimizer
toward less risky plans when less information is known about the filtered
values.

The pessimistic logic is triggered when the estimate derived from a histogram
is smaller than `table_row_count / 10,000`. This threshold is chosen because
we choose samples such that we expect to sample *nearly* every value with
multiplicity down to `table_row_count / 10,000` (see computeNumberSamples).
Selecivity estimates from a histogram below this resolution are suspect,
since there is increasing likelihood that a value was missed either due
to being omitted from the sample, or due to staleness.

Informs #130201

Release note (sql change): Added a clamp for row-count estimates over very
large tables so that the optimizer assumes that at least one distinct value
will be scanned. This reduces the chances of a catastrophic underestimate.
The new logic is off by default, gated by a session setting
`optimizer_clamp_low_histogram_selectivity`.

#### opt: use pessimistic estimates for inequality filters

This commit changes row count estimates for inequality filters, so that
we expect at least `rowCount / (bucketCount * 100)` rows to "survive" the
filter. This is in line with Postgres, which clamps inequality rowcount
estimates in a similar fashion.

Informs #130201

Release note (sql change): Added a clamp for the estimated selectivity of
inequality predicates that are unbounded on one or both sides (ex: `x > 5`).
This reduces the chances of a catastrophic understimate causing the optimizer
to choose a poorly-constrained scan. The new logic is off by default, gated by
the session setting `optimizer_clamp_inequality_selectivity`.

#### sql/opt: add telemetry counters for selectivity clamping

This commit adds telemetry counters for the new histogram selectivity
clamping behavior, as well as log messages to indicate when the new
behavior applies in a query's trace.

Epic: None

Release note: None

DrewKimball committed:
opt: make outside-of-histogram estimates more pessimistic

This commit makes rowcount estimation fall back on distinct count estimates
when a constraint includes zero histogram values. This biases the optimizer
toward less risky plans when less information is known about the filtered
values.

The pessimistic logic is triggered when the estimate derived from a histogram
is smaller than `table_row_count / 10,000`. This threshold is chosen because
we choose samples such that we expect to sample *nearly* every value with
multiplicity down to `table_row_count / 10,000` (see computeNumberSamples).
Selecivity estimates from a histogram below this resolution are suspect,
since there is increasing likelihood that a value was missed either due
to being omitted from the sample, or due to staleness.

Informs #130201

Release note (sql change): Added a clamp for row-count estimates over very
large tables so that the optimizer assumes that at least one distinct value
will be scanned. This reduces the chances of a catastrophic underestimate.
The new logic is off by default, gated by a session setting
`optimizer_clamp_low_histogram_selectivity`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/153067)
<!-- Reviewable:end -->

DrewKimball committed:
opt: use pessimistic estimates for inequality filters

This commit changes row count estimates for inequality filters, so that
we expect at least `rowCount / (bucketCount * 100)` rows to "survive" the
filter. This is in line with Postgres, which clamps inequality rowcount
estimates in a similar fashion.

Informs #130201

Release note (sql change): Added a clamp for the estimated selectivity of
inequality predicates that are unbounded on one or both sides (ex: `x > 5`).
This reduces the chances of a catastrophic understimate causing the optimizer
to choose a poorly-constrained scan. The new logic is off by default, gated by
the session setting `optimizer_clamp_inequality_selectivity`.

mw5h reviewed (changes requested):
> @mw5h reviewed 15 of 19 files at r3, 9 of 9 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153067)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`-- commits` line 27 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZufoUU1KKkyOHB466P:-OZufoUU1KKkyOHB466Q:b-7i29h7):*
nit: or due to staleness?
___
*[`pkg/sql/opt/memo/testdata/stats_quality/tpch/q03` line 181 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZum9Fs-tYumxjx0cTE:-OZum9Fs-tYumxjx0cTF:b-9dmcca) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34e3405c76c4cc6195410e074a73ca9e7b44ae1/pkg/sql/opt/memo/testdata/stats_quality/tpch/q03#L181)):*
> ```
> ----Stats for q3_lookup_join_4----
> WARNING: No previous statistics output was found. To collect actual statistics,
> run with the rewrite-actual-stats flag set to true.
> ```

Is this correct?
___
*[`pkg/sql/opt/xform/testdata/external/tpce` line 841 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZungeI1xE5IJFvcvKi:-OZungeI1xE5IJFvcvKj:btyzur4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34e3405c76c4cc6195410e074a73ca9e7b44ae1/pkg/sql/opt/xform/testdata/external/tpce#L841)):*
> ```
>  │    │    │    ├── columns: wi_wl_id:1!null wi_s_symb:2!null wl_c_id:6!null dm_date:10!null dm_s_symb:11!null dm_close:12!null lt_s_symb:18!null lt_price:20!null s_symb:25!null s_num_out:31!null
>  │    │    │    ├── fd: ()-->(1,6,10), (11)-->(12), (18)-->(20), (25)-->(31), (2)==(11,18,25), (11)==(2,18,25), (18)==(2,11,25), (25)==(2,11,18)
>  │    │    │    ├── inner-join (lookup daily_market)
> ```

Not really easy to determine if this plan is an improvement or not.
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 102 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZuh9C9D_z_klqyUw9H:-OZuh9C9D_z_klqyUw9I:b-kkkl46) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34e3405c76c4cc6195410e074a73ca9e7b44ae1/pkg/sql/opt/memo/statistics_builder.go#L102)):*
> ```Go
> 	// multiplicity >= row_count/10000. Cardinality estimates below this threshold
> 	// are increasingly likely to be inaccurate. See also computeNumberSamples.
> 	histogramPessimisticThreshold = 1.0 / 10000.0
> ```

Do we want to tie this to the actual values used in histogram collection?

Potential future work: Do we want to record these settings when we collect a histogram with the histogram?
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4571 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZuhnDD56ifw8Tjbf25:-OZuhnDD56ifw8Tjbf26:b-8717zv) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34e3405c76c4cc6195410e074a73ca9e7b44ae1/pkg/sql/opt/memo/statistics_builder.go#L4571)):*
> ```Go
> 			oldHist, newHist, inputColStat.DistinctCount, colStat.DistinctCount, inputStats.RowCount,
> 		)
> 		predicateSelectivity = props.MaxSelectivity(predicateSelectivity, pessimisticSel)
> ```

Can haz session variable switch?
___
*[`pkg/sql/opt/xform/testdata/external/tpcds/q71-q80` line 1277 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZuo63d0PH1EwJQw4iV:-OZuo63d0PH1EwJQw4iW:b-cz5m6w) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34e3405c76c4cc6195410e074a73ca9e7b44ae1/pkg/sql/opt/xform/testdata/external/tpcds/q71-q80#L1277)):*
> ```
>       │    │         ├── columns: channel:172!null col_name:173!null ws_ext_sales_price:105 item.i_category:130 date_dim.d_year:148 date_dim.d_qoy:152
>       │    │         ├── fd: ()-->(172,173)
>       │    │         ├── inner-join (hash)
> ```

This is concerning. Have you convinced yourself that this plan is at least a lateral move?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153067)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mw5h)

___
*[`pkg/sql/opt/memo/statistics_builder.go` line 102 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZuh9C9D_z_klqyUw9H:-OZusgKn0zN4hQPe6USN:bru1qbg) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34e3405c76c4cc6195410e074a73ca9e7b44ae1/pkg/sql/opt/memo/statistics_builder.go#L102)):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

Do we want to tie this to the actual values used in histogram collection?

Potential future work: Do we want to record these settings when we collect a histogram with the histogram?
</blockquote></details>

I considered that - mostly didn't do it because its a lot of annoying plumbing, and may be out-of-date anyway. I like the idea to record the sample size with the histogram.
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4571 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZuhnDD56ifw8Tjbf25:-OZurkCl9PzVFQaUyjVn:b8ad0l0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34e3405c76c4cc6195410e074a73ca9e7b44ae1/pkg/sql/opt/memo/statistics_builder.go#L4571)):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

Can haz session variable switch?
</blockquote></details>

Oh yes, I've been burned too many times by stats/coster changes with no switch.
___
*[`pkg/sql/opt/memo/testdata/stats_quality/tpch/q03` line 181 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZum9Fs-tYumxjx0cTE:-OZuq0esAKgCzseDj_oN:b-j00naw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34e3405c76c4cc6195410e074a73ca9e7b44ae1/pkg/sql/opt/memo/testdata/stats_quality/tpch/q03#L181)):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

Is this correct?
</blockquote></details>

I think I accidentally clobbered this test in a previous iteration where the plan changed - should be easy enough to just copy-paste the necessary stats back into place.

For context, these stats-quality tests rely on actual previously measured stats being kept around as part of the test case for comparison against the estimates. If the plan changes or you erase the execution stats, you get the error message above.
___
*[`pkg/sql/opt/xform/testdata/external/tpce` line 841 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZungeI1xE5IJFvcvKi:-OZuqFDzCeHwhEpbmrDf:b-wi2u53) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34e3405c76c4cc6195410e074a73ca9e7b44ae1/pkg/sql/opt/xform/testdata/external/tpce#L841)):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

Not really easy to determine if this plan is an improvement or not.
</blockquote></details>

IIRC, the most important thing with these queries was just to get the lookup-join pipeline, but yeah, will have to actually run it to see for sure.
___
*[`pkg/sql/opt/xform/testdata/external/tpcds/q71-q80` line 1277 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZuo63d0PH1EwJQw4iV:-OZurLJq5UxCX39RXjNX:b-8jv3rp) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34e3405c76c4cc6195410e074a73ca9e7b44ae1/pkg/sql/opt/xform/testdata/external/tpcds/q71-q80#L1277)):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

This is concerning. Have you convinced yourself that this plan is at least a lateral move?
</blockquote></details>

I'll admit I skimmed past this one... will investigate it more deeply. That being said, if any workload is going to use a hash join instead of a lookup join, TPC-DS seems like a good candidate.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153067)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mw5h)

___
*[`-- commits` line 27 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZufoUU1KKkyOHB466P:-OZuz9AK7t994150pek3:b-896fix):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

nit: or due to staleness?
</blockquote></details>

Done.
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4571 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZuhnDD56ifw8Tjbf25:-OZuzAYoE2glO2R5mq8R:bm7679) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34e3405c76c4cc6195410e074a73ca9e7b44ae1/pkg/sql/opt/memo/statistics_builder.go#L4571)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Oh yes, I've been burned too many times by stats/coster changes with no switch.
</blockquote></details>

Added a TODO, will add one pre-merge if we decide to go forward with it.
___
*[`pkg/sql/opt/xform/testdata/external/tpcds/q71-q80` line 1277 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZuo63d0PH1EwJQw4iV:-OZuyKjT2WX6yE7rS_og:b-8dykj8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34e3405c76c4cc6195410e074a73ca9e7b44ae1/pkg/sql/opt/xform/testdata/external/tpcds/q71-q80#L1277)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I'll admit I skimmed past this one... will investigate it more deeply. That being said, if any workload is going to use a hash join instead of a lookup join, TPC-DS seems like a good candidate.
</blockquote></details>

I tried out both plans. The plan doesn't have an effect on my machine at least, since all the time is spent elsewhere in larger scans.


<!-- Sent from Reviewable.io -->

mw5h reviewed (approved):
I think this looks great, aside from needing an off switch. Because the changes are so localized, I feel like you could backport them safely.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @mw5h reviewed 7 of 11 files at r6, 9 of 9 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-:-OZvEHKJ2KfedlU1l-5j:bvdmsrw)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4590 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZvDsy1BnknHmvrJjOl:-OZvDsy1BnknHmvrJjOm:b9d359o) ([raw file](https://github.com/cockroachdb/cockroach/blob/50993410537a62df3b59bb03c7af330f6bfa81e7/pkg/sql/opt/memo/statistics_builder.go#L4590)):*
> ```Go
> 	oldHist, newHist *props.Histogram, oldDistinct, newDistinct, tableRowCount float64,
> ) props.Selectivity {
> 	// TODO(drewk): add a version gate here.
> ```

nit: switch, not version gate.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
This is promising, but we might want to hold off until 25.4 ships so that we can do more testing.

> @michae2 reviewed 18 of 18 files at r2, 16 of 16 files at r5, 1 of 11 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-:-OZz41wH0gAdeWT_UriG:b-t6ta09)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/opt/xform/testdata/coster/outside-histogram` line 94 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZyo6uJ36RJW8yy6rHP:-OZyo6uJ36RJW8yy6rHQ:b-4z40c6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8eed40a8f43d6e563bd614cdd5c990cfbf6d00b/pkg/sql/opt/xform/testdata/coster/outside-histogram#L94)):*
> ```
>       {"num_eq": 1000000, "num_range": 199000000, "distinct_range": 199000000, "upper_bound": "200000000"},
>       {"num_eq": 1000000, "num_range": 299000000, "distinct_range": 299000000, "upper_bound": "300000000"},
>       {"num_eq": 1000000, "num_range": 399000000, "distinct_range": 399000000, "upper_bound": "400000000"}
> ```

If k is the primary key, num_eq should be at most 1.

Also, there isn't room for 399m distinct integers between 300m and 400m.
___
*[`pkg/sql/opt/xform/testdata/coster/outside-histogram` line 129 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZypQR73eSq8BNdcF-I:-OZypQR73eSq8BNdcF-J:b-nybkj0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8eed40a8f43d6e563bd614cdd5c990cfbf6d00b/pkg/sql/opt/xform/testdata/coster/outside-histogram#L129)):*
> ```
>     ]
>   }
> ]'
> ```

There's another source of staleness that would also be good to check: an enum column with a value missing from the histogram (and all buckets in the histogram with num_range=0).
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4598 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZz2-rf539ELhMh-Tdp:-OZz2-rf539ELhMh-Tdq:b-s54lgb) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc9b8c936041869cbed96a23903d98047c374a82/pkg/sql/opt/memo/statistics_builder.go#L4598)):*
> ```Go
> 		// counts in selectivityFromSingleColDistinctCounts, so this doesn't
> 		// double count the effect of the predicate.
> 		pessimisticSel := props.MakeSelectivityFromFraction(newDistinct, oldDistinct)
> ```

I'm a little confused about using the distinct counts here.

If we suspect that newHistValues is an undercount, then it seems like newDistinct could also be an undercount?
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4609 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZz35A04uoqTeo_753x:-OZz35A04uoqTeo_753y:btdlvhv) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc9b8c936041869cbed96a23903d98047c374a82/pkg/sql/opt/memo/statistics_builder.go#L4609)):*
> ```Go
> 		// selectivity, since an inverted index can have more rows than the primary
> 		// key.
> 		inputRowCount := max(tableRowCount, oldHistValues)
> ```

Is oldHistValues right here? It seems like if we're filtering an inverted index, this should be the total count of values in the "whole" inverted histogram.

And if we're filtering a forward index or a regular column, it should also be the total count of values in the "whole" column histogram.

I'm not sure we have access to the "whole" histogram at this point, but if we do, maybe this should simply be ValuesCount() of the whole histogram instead of using max?


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> @mgartner reviewed 18 of 18 files at r2, 16 of 16 files at r5, 11 of 11 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153067)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @michae2)

___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4609 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZz35A04uoqTeo_753x:-O_CokL0CJwmOHpo9_uF:bh6d5ff) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc9b8c936041869cbed96a23903d98047c374a82/pkg/sql/opt/memo/statistics_builder.go#L4609)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Is oldHistValues right here? It seems like if we're filtering an inverted index, this should be the total count of values in the "whole" inverted histogram.

And if we're filtering a forward index or a regular column, it should also be the total count of values in the "whole" column histogram.

I'm not sure we have access to the "whole" histogram at this point, but if we do, maybe this should simply be ValuesCount() of the whole histogram instead of using max?
</blockquote></details>

I think this is logically correct, but `tableRowCount` is confusingly named. It should be in terms of input, not table, since the input might not be a table scan.
___
*[`pkg/sql/opt/xform/testdata/coster/outside-histogram` line 900 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-O_CkWO-7EDnW5xv192Y:-O_CkWO-7EDnW5xv192Z:bff1yho) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8eed40a8f43d6e563bd614cdd5c990cfbf6d00b/pkg/sql/opt/xform/testdata/coster/outside-histogram#L900)):*
> ```
> 
> opt set=(optimizer_prefer_bounded_cardinality=false,optimizer_min_row_count=0)
> SELECT * FROM t_large WHERE k = 100000000 AND i = 500000000 AND s = 'zzz'
> ```

I organized these tests where there is a `Qn` section for each query, and it is run with all combinations of `optimizer_prefer_bounded_cardinality` and `optimizer_in_row_count=0/1`. Do you mind following this pattern for the new queries you added?
___
*[`pkg/sql/opt/exec/execbuilder/testdata/forecast1401` line 18408 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-O_Cp-422T3V8hYq-wq9:-O_Cp-422T3V8hYq-wqA:bgtud9p) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc9b8c936041869cbed96a23903d98047c374a82/pkg/sql/opt/exec/execbuilder/testdata/forecast1401#L18408)):*
> ```
>   spans: [/'2022-01-25 21:17:27.216095+00' - ]
> 
> # Without forecasts, the bad plan uses t1401_c_idx.
> ```

nit: add a test like this with this new logic disabled so we see the bad plan still.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/forecast1401` line 18458 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-O_Cp7ZH4oILSxZXZkDJ:-O_Cp7ZH4oILSxZXZkDK:bvbd4fi) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc9b8c936041869cbed96a23903d98047c374a82/pkg/sql/opt/exec/execbuilder/testdata/forecast1401#L18458)):*
> ```
>   spans: [/'2022-01-25 21:17:27.216095+00' - ]
> 
> # With forecasts, the good plan uses t1401_f_c_idx1.
> ```

nit: ditto ^


<!-- Sent from Reviewable.io -->

DrewKimball committed:
opt: add outside-histogram tests with large table

This commit adds to the existing optimizer tests for filters that select
values outside of histograms. The new tests are against a large table
to show how the optimizer's trust of low-selectivity estimates varies
with table size.

Epic: None

Release note: None

DrewKimball committed:
opt: improve outside-histogram optimizer tests

This commit makes the following improvements to the `outside-histogram`
optimizer test:
* Fixes a minor mistake in the test, where the upper bounds on a histogram
  for a unique column were incorrect.
* Simplifies the test code for setting session variable defaults.
* Adds comments explaining each test case.

Epic: None

Release note: None

DrewKimball committed:
opt: add outside-histogram test for missing enum value

This commit adds a pair of `outside-histogram` optimizer test cases that
simulate a query against an enum column for which one of the values was not
sampled.

Epic: None

Release note: None

DrewKimball commented on pkg/sql/opt/memo/statistics_builder.go:
```diff
@@ -90,6 +90,16 @@ const (
+	// to table size such that we expect to *nearly* always sample all values with
+	// multiplicity >= row_count/10000. Cardinality estimates below this threshold
+	// are increasingly likely to be inaccurate. See also computeNumberSamples.
+	histogramPessimisticThreshold = 1.0 / 10000.0
```
I'd be inclined to keep the knob surface area smaller, but open to push-back on that.
<!-- Reviewable comment -O_VIFuh2BcVG_2H-kGk:bikuxzr -->
<!-- Sent from Reviewable.io -->

DrewKimball commented on pkg/sql/opt/memo/statistics_builder.go:
```diff
@@ -4562,6 +4580,54 @@ func (sb *statisticsBuilder) selectivityFromHistograms(
+	// Calculate the expected number of out-of-histogram values that will pass
+	// through the filters.
+	pessimisticThreshold := tableRowCount * histogramPessimisticThreshold
+	if newHistValues < pessimisticThreshold {
```
Do you mean a telemetry counter? I think if we do that, we should probably make sure it's only incremented once per query.
<!-- Reviewable comment -O_VI428CP1wVa11SmXH:bqc2dpb -->
<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
Thanks everyone for taking a look! I ended up re-structuring the logic that clamps the selectivity for very low estimates, since I think the first version was difficult to understand/justify. Now, the "resolution" of the histogram is stored on the histogram itself and set when the histogram is built. Also, I've added two session variables to gate the two forms of "distrust" we're introducing, both off by default.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-:-O_VJ6boAL2EfvsM7uFi:b-69zjqb)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @michae2, and @mw5h)

___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4598 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZz2-rf539ELhMh-Tdp:-O_OmgOTE9daIrQ4d2OF:bz6ry6) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc9b8c936041869cbed96a23903d98047c374a82/pkg/sql/opt/memo/statistics_builder.go#L4598)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I'm a little confused about using the distinct counts here.

If we suspect that newHistValues is an undercount, then it seems like newDistinct could also be an undercount?
</blockquote></details>

We're already a bit pessimistic about how distinct counts are updated: https://github.com/cockroachdb/cockroach/blob/ba4fefb34695f38481ce218e6faede273ef4f8b7/pkg/sql/opt/memo/statistics_builder.go#L3890-L3894

And we're careful when updating distinct counts from histograms, so that it isn't just a straightforward reduction by `selectivity` when the distinct count is small relative to the row count: https://github.com/cockroachdb/cockroach/blob/ba4fefb34695f38481ce218e6faede273ef4f8b7/pkg/sql/opt/props/histogram.go#L96-L103
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4609 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZz35A04uoqTeo_753x:-O_J3H6z2g6BRoVteMQf:b-7b6asv) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc9b8c936041869cbed96a23903d98047c374a82/pkg/sql/opt/memo/statistics_builder.go#L4609)):*
>I'm not sure we have access to the "whole" histogram at this point, but if we do, maybe this should simply be ValuesCount() of the whole histogram instead of using max?

That should be what `oldHistValues` is - it's defined as `oldHist.ValuesCount()`, which sums up all the values contained in the histogram.
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4590 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZvDsy1BnknHmvrJjOl:-O_VIEz96GwEeUuz7XD5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/50993410537a62df3b59bb03c7af330f6bfa81e7/pkg/sql/opt/memo/statistics_builder.go#L4590)):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

nit: switch, not version gate.
</blockquote></details>

Done.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/forecast1401` line 18408 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-O_Cp-422T3V8hYq-wq9:-O_VIeua2jB0EBLouMQl:b5m2sov) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc9b8c936041869cbed96a23903d98047c374a82/pkg/sql/opt/exec/execbuilder/testdata/forecast1401#L18408)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: add a test like this with this new logic disabled so we see the bad plan still.
</blockquote></details>

With the session variables, this test is no longer changed.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/forecast1401` line 18458 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-O_Cp7ZH4oILSxZXZkDJ:-O_VIr1V6NcmzRCjoTiL:b5m2sov) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc9b8c936041869cbed96a23903d98047c374a82/pkg/sql/opt/exec/execbuilder/testdata/forecast1401#L18458)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: ditto ^
</blockquote></details>

With the session variables, this test is no longer changed.
___
*[`pkg/sql/opt/xform/testdata/coster/outside-histogram` line 94 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZyo6uJ36RJW8yy6rHP:-O_FubXO20Mw_PbrwjcT:b9tqfln) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8eed40a8f43d6e563bd614cdd5c990cfbf6d00b/pkg/sql/opt/xform/testdata/coster/outside-histogram#L94)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

If k is the primary key, num_eq should be at most 1.

Also, there isn't room for 399m distinct integers between 300m and 400m.
</blockquote></details>

Good catch, I was lazy. The original test had a similar issue, so I added a commit to fix that as well.
___
*[`pkg/sql/opt/xform/testdata/coster/outside-histogram` line 129 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZypQR73eSq8BNdcF-I:-O_FwYvyCdeK1Vavxjie:beogqr3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8eed40a8f43d6e563bd614cdd5c990cfbf6d00b/pkg/sql/opt/xform/testdata/coster/outside-histogram#L129)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

There's another source of staleness that would also be good to check: an enum column with a value missing from the histogram (and all buckets in the histogram with num_range=0).
</blockquote></details>

That's a good test idea, done.
___
*[`pkg/sql/opt/xform/testdata/coster/outside-histogram` line 900 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-O_CkWO-7EDnW5xv192Y:-O_VIs5k09oOvyh6Ora6:b-9ogfkk) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8eed40a8f43d6e563bd614cdd5c990cfbf6d00b/pkg/sql/opt/xform/testdata/coster/outside-histogram#L900)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I organized these tests where there is a `Qn` section for each query, and it is run with all combinations of `optimizer_prefer_bounded_cardinality` and `optimizer_in_row_count=0/1`. Do you mind following this pattern for the new queries you added?
</blockquote></details>

I went with `Qn (large)` for these cases with the larger table, and added the new session variable options. Is that approximately what you had in mind?


<!-- Sent from Reviewable.io -->

mw5h commented:
> Do you mean a telemetry counter? I think if we do that, we should probably make sure it's only incremented once per query.

That's true and I guess makes it harder, but I think knowing when we're falling off of pure stat usage to backup heuristics is helpful in debugging and maybe even determining when stats are stale (depending on resolution). Could we add a couple of flags to the memo and then increment the counter based on those?

DrewKimball committed:
sql/opt: add telemetry counters for selectivity clamping

This commit adds telemetry counters for the new histogram selectivity
clamping behavior, as well as log messages to indicate when the new
behavior applies in a query's trace.

Epic: None

Release note: None

DrewKimball commented:
I've added one more commit that increments a pair of telemetry counters and writes a log message when these selectivity clamps apply during planning.

mw5h reviewed (approved):
Thanks! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work! I had one important question re: the inequality clamping.

> @mgartner reviewed 3 of 19 files at r8, 16 of 16 files at r19, 1 of 1 files at r20, 1 of 1 files at r21, 13 of 13 files at r22, 11 of 11 files at r23, 5 of 5 files at r24, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-:-OauUshMCeokFDDABxkd:b-spdlgg)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @michae2, and @mw5h)

___
*[`pkg/sql/opt/xform/testdata/coster/outside-histogram` line 900 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-O_CkWO-7EDnW5xv192Y:-OauLK6t79SeB-WOQdYH:bf39w59) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8eed40a8f43d6e563bd614cdd5c990cfbf6d00b/pkg/sql/opt/xform/testdata/coster/outside-histogram#L900)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I went with `Qn (large)` for these cases with the larger table, and added the new session variable options. Is that approximately what you had in mind?
</blockquote></details>

looks great! thanks!
___
*[`-- commits` line 53 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OauOzds5foyRUBMNugz:-OauOzds5foyRUBMNuh-:b-2l97mt):*
nit: Should the knob be documented in the release note?
___
*[`pkg/sql/opt/xform/testdata/coster/outside-histogram` line 2476 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OauTuN1-N3-O41CnJwZ:-OauTuN1-N3-O41CnJw_:b-drsa3o) ([raw file](https://github.com/cockroachdb/cockroach/blob/7eaa39f9b160dd0fb086e41414bdec63368b9e59/pkg/sql/opt/xform/testdata/coster/outside-histogram#L2476)):*
> ```
>       └── e:4 = 'PENDING' [outer=(4), immutable, constraints=(/4: [/'PENDING' - /'PENDING']; tight), fd=()-->(4)]
> 
> opt set=optimizer_clamp_inequality_selectivity=true
> ```

nit: I'd argue it's not necessary to tests queries without inequalities with this setting, but up to you. 
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4624 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OauT6TL99etxnOW82P_:-OauT6TL99etxnOW82Pa:bwybngm) ([raw file](https://github.com/cockroachdb/cockroach/blob/7eaa39f9b160dd0fb086e41414bdec63368b9e59/pkg/sql/opt/memo/statistics_builder.go#L4624)):*
> ```Go
> 		(!tightUpperBound || !tightLowerBound) {
> 		// Similar to Postgres, ratchet the selectivity estimate towards 1/100th
> 		// of the average histogram bucket size if the filters do not bound the
> ```

nit: "average histogram bucket size" sounds like "average number of rows per bucket" not "number of buckets", which `BucketCount()` returns. Which is it supposed to be? The former makes more sense to me, but that's not what the code is doing, unless I'm missing something.


<!-- Sent from Reviewable.io -->

DrewKimball commented on pkg/sql/opt/memo/statistics_builder.go:
```diff
@@ -4562,6 +4580,54 @@ func (sb *statisticsBuilder) selectivityFromHistograms(
+	}
+
+	tightUpperBound, tightLowerBound := newHist.TightBounds()
+	if !tightUpperBound || !tightLowerBound {
```
Done.
<!-- Reviewable comment -OauZzu8EPj2-mzzaLq-:b-896fix -->
<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-:-OaubAud7sCgjy8MXElE:bz5mair)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @michae2, and @mw5h)

___
*[`-- commits` line 53 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OauOzds5foyRUBMNugz:-OauZy8_430Aqc25pXez:b-gh93gd):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: Should the knob be documented in the release note?
</blockquote></details>

Good idea, Done.
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4624 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OauT6TL99etxnOW82P_:-OauXB_E7wJFL0hx-Ton:btwbnth) ([raw file](https://github.com/cockroachdb/cockroach/blob/7eaa39f9b160dd0fb086e41414bdec63368b9e59/pkg/sql/opt/memo/statistics_builder.go#L4624)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: "average histogram bucket size" sounds like "average number of rows per bucket" not "number of buckets", which `BucketCount()` returns. Which is it supposed to be? The former makes more sense to me, but that's not what the code is doing, unless I'm missing something.
</blockquote></details>

We're clamping the row count estimate to 1/100th of the average bucket size, and the selectivity to `1/(100*bucket_count)`. Those things are equivalent, but the comment is unclear.

I've changed the comment to say this:
```
// Similar to Postgres, assume that an open-ended inequality predicate will
// scan at least 1/100th of the average number of rows in a histogram
// bucket. This accounts for the possibility that the histogram missed
// extreme values due to sampling or staleness.
```
___
*[`pkg/sql/opt/xform/testdata/coster/outside-histogram` line 2476 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OauTuN1-N3-O41CnJwZ:-OauXv-H1zlVyQDLB63m:b7qfjhu) ([raw file](https://github.com/cockroachdb/cockroach/blob/7eaa39f9b160dd0fb086e41414bdec63368b9e59/pkg/sql/opt/xform/testdata/coster/outside-histogram#L2476)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: I'd argue it's not necessary to tests queries without inequalities with this setting, but up to you.
</blockquote></details>

I think it's still useful to show when the heuristics don't apply, so I'd prefer to keep these.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> @mgartner reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153067)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @DrewKimball, @michae2, and @mw5h)

___
*[`-- commits` line 53 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OauOzds5foyRUBMNugz:-OaufLAl5cXUO27hny4n:b-jtzzv3):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Good idea, Done.
</blockquote></details>

Don't forget to update the PR description.
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4624 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OauT6TL99etxnOW82P_:-OaufCPA2itnBAsVUZBo:b-8foaw1) ([raw file](https://github.com/cockroachdb/cockroach/blob/7eaa39f9b160dd0fb086e41414bdec63368b9e59/pkg/sql/opt/memo/statistics_builder.go#L4624)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

We're clamping the row count estimate to 1/100th of the average bucket size, and the selectivity to `1/(100*bucket_count)`. Those things are equivalent, but the comment is unclear.

I've changed the comment to say this:
```
// Similar to Postgres, assume that an open-ended inequality predicate will
// scan at least 1/100th of the average number of rows in a histogram
// bucket. This accounts for the possibility that the histogram missed
// extreme values due to sampling or staleness.
```
</blockquote></details>

Oh right, duh. LGTM.
___
*[`pkg/sql/opt/xform/testdata/coster/outside-histogram` line 2476 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OauTuN1-N3-O41CnJwZ:-OaufFH9-Or37zKaJr0d:b-81nlyp) ([raw file](https://github.com/cockroachdb/cockroach/blob/7eaa39f9b160dd0fb086e41414bdec63368b9e59/pkg/sql/opt/xform/testdata/coster/outside-histogram#L2476)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I think it's still useful to show when the heuristics don't apply, so I'd prefer to keep these.
</blockquote></details>

SGTM!


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> This is exciting!

> @michae2 reviewed 8 of 19 files at r3, 3 of 19 files at r8, 13 of 24 files at r12, 16 of 16 files at r31, 1 of 1 files at r32, 1 of 1 files at r33, 13 of 13 files at r34, 11 of 11 files at r35, 5 of 5 files at r36, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-:-Oax9vmK-AuzmmU7kKzZ:b-hm5c4p)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @DrewKimball and @mw5h)

___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4598 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZz2-rf539ELhMh-Tdp:-Oax4a5m5f8wVXzKF0Bw:b-s2bugm) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc9b8c936041869cbed96a23903d98047c374a82/pkg/sql/opt/memo/statistics_builder.go#L4598)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

We're already a bit pessimistic about how distinct counts are updated: https://github.com/cockroachdb/cockroach/blob/ba4fefb34695f38481ce218e6faede273ef4f8b7/pkg/sql/opt/memo/statistics_builder.go#L3890-L3894

And we're careful when updating distinct counts from histograms, so that it isn't just a straightforward reduction by `selectivity` when the distinct count is small relative to the row count: https://github.com/cockroachdb/cockroach/blob/ba4fefb34695f38481ce218e6faede273ef4f8b7/pkg/sql/opt/props/histogram.go#L96-L103
</blockquote></details>

Ok, thanks for explaining.
___
*[`pkg/sql/opt/xform/testdata/coster/outside-histogram` line 94 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OZyo6uJ36RJW8yy6rHP:-Oax9G-GB6OtV6nSgMEz:b-9755fn) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8eed40a8f43d6e563bd614cdd5c990cfbf6d00b/pkg/sql/opt/xform/testdata/coster/outside-histogram#L94)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Good catch, I was lazy. The original test had a similar issue, so I added a commit to fix that as well.
</blockquote></details>

Thank you!
___
*[`pkg/sql/opt/testutils/testcat/set_var.go` line 15 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OawhaKY51tLjBRFDMo1:-OawhaKY51tLjBRFDMo2:b-174v7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce3256e3bc39213c618f123115cec3555130d385/pkg/sql/opt/testutils/testcat/set_var.go#L15)):*
> ```Go
> 
> // SetVar implements the 'SET ...' SQL statement. Currently, it supports only
> // statements that set the current role (e.g., SET ROLE <user>).
> ```

I think the second sentence of this comment can be removed now.
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4627 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-Oax78pFElSp3RyQ5tlW:-Oax78pFElSp3RyQ5tlX:b-rwv8el) ([raw file](https://github.com/cockroachdb/cockroach/blob/1228ab99587e2e84b939341848942d70e11dddb8/pkg/sql/opt/memo/statistics_builder.go#L4627)):*
> ```Go
> 		// bucket. This accounts for the possibility that the histogram missed
> 		// extreme values due to sampling or staleness.
> 		inequalityClamp := props.MakeSelectivityFromFraction(1, float64(oldHist.BucketCount())*100)
> ```

I'm a little hesitant to use bucket count in this calculation. The number of histogram buckets could be influenced by the user (via `sql_stats_histogram_buckets_count`), or by how many partial stats we've merged, or by how many full stats were included in a forecast. It's somewhat arbitrary. Maybe bucket count has more significance in PG? (Maybe it's fixed per table?)

Instead of bucket count, could we use some fixed fraction, like `histogramPessimisticThreshold`?
___
*[`pkg/sql/opt/xform/testdata/coster/outside-histogram` line 1299 at r33](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OawpxJI-lAOPNzQeozK:-OawpxJI-lAOPNzQeozL:bqtt704) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e29a381b7af3ad546ed7a397ce95eb1da1bbc2a/pkg/sql/opt/xform/testdata/coster/outside-histogram#L1299)):*
> ```
>  ├── key: ()
>  ├── fd: ()-->(1-4)
>  ├── scan t_large
> ```

Funny, I was expecting the large table to run into problems where optimizer_min_row_count=1 doesn't help due to the table size. But this case seems to show the opposite: the larger size table uses the primary index even without optimizer_min_row_count=1.

Is there a case that shows optimizer_min_row_count=1 not helping the large table? Maybe we need 10x the number of rows in the large table? Or more keys in the IN list?


<!-- Sent from Reviewable.io -->

DrewKimball committed:
opt: add unique and distinct column to outside-histogram test

This commit adds a column `u` with a unique index, and a column `d`
with a non-unique index to the outside-histogram test table. Both
columns use the same histogram as the primary key `k` (and therefore
have no duplicate values).

This commit also adds tests that are similar to `Q1`, but using the
new columns instead of `k`. This shows the optimizer's behavior when
the "good" plan requires an index join, and when filtering on `d`,
shows how the optimizer behaves when no plan limits the max cardinality
of the query.

Epic: None

Release note: None

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153067)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @michae2 and @mw5h)

___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4627 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-Oax78pFElSp3RyQ5tlW:-OcEBRZnBTNm-0zmpRpF:b3ocylj) ([raw file](https://github.com/cockroachdb/cockroach/blob/1228ab99587e2e84b939341848942d70e11dddb8/pkg/sql/opt/memo/statistics_builder.go#L4627)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I'm a little hesitant to use bucket count in this calculation. The number of histogram buckets could be influenced by the user (via `sql_stats_histogram_buckets_count`), or by how many partial stats we've merged, or by how many full stats were included in a forecast. It's somewhat arbitrary. Maybe bucket count has more significance in PG? (Maybe it's fixed per table?)

Instead of bucket count, could we use some fixed fraction, like `histogramPessimisticThreshold`?
</blockquote></details>

Done. PG uses 100 histograms by default, and this isn't far off from our default (200), so I went with `1/10000` as the constant.
___
*[`pkg/sql/opt/testutils/testcat/set_var.go` line 15 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OawhaKY51tLjBRFDMo1:-Ob41ai_6iMGeEsrQKs4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce3256e3bc39213c618f123115cec3555130d385/pkg/sql/opt/testutils/testcat/set_var.go#L15)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I think the second sentence of this comment can be removed now.
</blockquote></details>

Done.
___
*[`pkg/sql/opt/xform/testdata/coster/outside-histogram` line 1299 at r33](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OawpxJI-lAOPNzQeozK:-OcEA8Wd7qthDomHL5Eb:bwrnoxi) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e29a381b7af3ad546ed7a397ce95eb1da1bbc2a/pkg/sql/opt/xform/testdata/coster/outside-histogram#L1299)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Funny, I was expecting the large table to run into problems where optimizer_min_row_count=1 doesn't help due to the table size. But this case seems to show the opposite: the larger size table uses the primary index even without optimizer_min_row_count=1.

Is there a case that shows optimizer_min_row_count=1 not helping the large table? Maybe we need 10x the number of rows in the large table? Or more keys in the IN list?
</blockquote></details>

That's a good point. I added a commit that adds a unique column (with no stored columns), as well as a non-unique indexed column, both of which use the same histogram as "k" (e.g. no duplicates). Tests 7-10 now use these columns. `min_row_count` isn't enough to help in these cases because there is now an index join in both the good and bad plans, while `prefer_bounded_cardinality` doesn't help for the non-unique column.


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @michae2 reviewed 17 of 17 files at r37, 1 of 1 files at r38, 1 of 1 files at r39, 1 of 1 files at r40, 11 of 13 files at r41.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-:-OcfRDUK4LCXcMtX3fWl:bnfp4nl)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @mw5h)

___
*[`pkg/sql/opt/xform/testdata/coster/outside-histogram` line 1299 at r33](https://reviewable.io/reviews/cockroachdb/cockroach/153067#-OawpxJI-lAOPNzQeozK:-OcfRR1a6W6lsQaN9iO8:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e29a381b7af3ad546ed7a397ce95eb1da1bbc2a/pkg/sql/opt/xform/testdata/coster/outside-histogram#L1299)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

That's a good point. I added a commit that adds a unique column (with no stored columns), as well as a non-unique indexed column, both of which use the same histogram as "k" (e.g. no duplicates). Tests 7-10 now use these columns. `min_row_count` isn't enough to help in these cases because there is now an index join in both the good and bad plans, while `prefer_bounded_cardinality` doesn't help for the non-unique column.
</blockquote></details>

Thanks!


<!-- Sent from Reviewable.io -->
